### PR TITLE
SYNTHESIZER: Migrate loop-contracts synthesizer from goto-instrument into goto-synthesizer

### PR DESCRIPTION
**Pull request by qinheping**
This PR copies the synthesizer interface and expression enumerator from `goto-instrument` into `goto-synthesizer`, and removes them from `goto-instrument.

### CODE REVIEW

File: src/goto-synthesizer/Makefile [hunk 1]
```diff
@@ -1,6 +1,9 @@
-SRC = goto_synthesizer_languages.cpp \
+SRC = enumerative_loop_contracts_synthesizer.cpp \
+      expr_enumerator.cpp \
+      goto_synthesizer_languages.cpp \
       goto_synthesizer_main.cpp \
       goto_synthesizer_parse_options.cpp \
+      synthesizer.utils.cpp \
```
Comment (tautschnig) [inline src/goto-synthesizer/Makefile]: Is this the actual file name?

File: src/goto-instrument/havoc_utils.cpp [hunk 2]
```diff
@@ -16,7 +16,14 @@ Date: July 2021
 #include <util/pointer_expr.h>
 #include <util/std_code.h>
 
-#include <goto-programs/goto_program.h>
+#define IS_LOOP_HAVOC "Loop havocing instrumented by applying loop contracts."
```
Comment (tautschnig) [inline src/goto-instrument/havoc_utils.cpp]: Why is new code being added here?

File: unit/CMakeLists.txt [hunk 3]
```diff
@@ -84,6 +84,7 @@ target_link_libraries(
         goto-checker
         goto-programs
         goto-instrument-lib
+        goto-synthesizer-lib
```
Comment (tautschnig) [inline unit/CMakeLists.txt:87]: This really suggests that you might want to fold the two PRs into the same commit.

File: src/goto-synthesizer/Makefile (hunk 1 above)
Comment (qinheping) [reply src/goto-synthesizer/Makefile]: It should be `synthesizer_utils`. Thank you.

File: src/goto-instrument/havoc_utils.cpp (hunk 2 above)
Comment (qinheping) [reply src/goto-instrument/havoc_utils.cpp]: They should be added in the other PR. I am going to remove these new codes from `havoc_utils.cpp` and `havoc_utils.h`.

File: unit/CMakeLists.txt (hunk 3 above)
Comment (qinheping) [reply unit/CMakeLists.txt:87]: Good point. I combined them to one PR now.

**Review by tautschnig — APPROVED** [review]
I believe the man pages are also to be updated, and the same is true of the `CODEOWNERS` file.

**Pull request merged**